rust-spec/lightclient/detection/detection.md

```diff
@@ -276,421 +447,352 @@ This implies that we need a timeout of at least *2 Delta* for remote
-            - **FUTURE:** try to do fork detection from some possibly old
-             lightblock in store. (Might be the approach for the
-             light node that assumes to be connected to correct
-             full nodes only from time to time)
+```go
+func AttackDetector(root LightBlock, primary_trace []LightBlock) 
+                   ([]InternalEvidence) {
+   
+    Evidences := new []InternalEvidence;
+ 
+    for each secondary in Secondaries {
+        // we replay the primary trace with the secondary, in
+        // order to generate evidence that we can submit to the
+        // secodary. We return the evidence + the trace the
+        // secondary told us that spans the evidence at its local store
+   
+        EvidenceForSecondary, newroot, secondary_trace, result :=
+                CreateEvidenceForPeer(secondary, 
+				                      root, 
+									  primary_trace);
+        if result == FaultyPeer {
+            Replace_Secondary(root);
+        }
+        else if result == FoundEvidence {
+            // the conflict is not bogus
+            Evidences.Add(EvidenceForSecondary);
+            // we replay the secondary trace with the primary, ...
+            EvidenceForPrimary, _, result := 
+				    CreateEvidenceForPeer(primary,
+                                          newroot,
+                                          secondary_trace);
+            if result == FoundEvidence {
+                Evidences.Add(EvidenceForPrimary);
+            }
+            // At this point we do not care about the other error
+            // codes. We already have generated evidence for an
+            // attack and need to stop the lightclient. It does not
+            // help to call replace_primary. Also we will use the
+            // same primary to check with other secondaries in
+            // later iterations of the loop
+         }
+         // In the case where the secondary reports NoEvidence
+         // we do nothing
+    }
+    return Evidences; 
+}
+```
 
 - Expected precondition
-    - *FullNodes* is nonempty
+    - root and primary trace are a verification trace
 - Expected postcondition
-    - addr is moved from *Secondaries* to *FaultyNodes*
-    - an address *a* is moved from *FullNodes* to *Secondaries*
+    - solves the problem statement (if attack found, then evidence is reported)
 - Error condition
-    - if precondition is violated
+    - `ErrorTrustExpired`: fails if root expires (outside trusting
+    period) [[LCV-INV-TP.1]](LCV-INV-TP1-link)
+    - `ErrorNoPeers`: if no peers are left to replace secondaries, and
+    no evidence was found before that happened
 
-### From the verifier
+----
 
 ```go
-func VerifyToTarget(primary PeerID, lightStore LightStore,
-                    targetHeight Height) (LightStore, Result)
+func CreateEvidenceForPeer(peer PeerID, root LightBlock, trace LightStore)
+                          (Evidence, LightStore, result) {
```

Comment:
I think we are missing one return value as it returns 4 values.

Reply:
yes, I added the LightBlock at the second position